File: android/src/main/java/hu/hauserresults/plugins/caprealtime/CapacitorRealTimePlugin.java
//CapacitorRealTimePlugin.java
package hu.hauserresults.plugins.caprealtime;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.net.InetAddress;
import java.time.Clock;
import org.apache.commons.net.ntp.NTPUDPClient;
import org.apache.commons.net.ntp.TimeInfo;

@CapacitorPlugin(name = "CapacitorRealTime")
public class CapacitorRealTimePlugin extends Plugin {

    private CapacitorRealTime implementation = new CapacitorRealTime();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void getUptime(PluginCall call) {
        long uptime = implementation.getUptime();
        JSObject ret = new JSObject();
        ret.put("uptime", uptime);
        call.resolve(ret);
    }

    @PluginMethod
    public void getGnssTime(PluginCall call) {
        Clock gnssTime = implementation.getGnssTime();
        JSObject ret = new JSObject();

        if(gnssTime === -1) {
            call.reject("Error getting GNSS Time");
        }else if(gnssTime === null) {
            call.reject("GNSS Time is not available");
        }else{
            ret.put("gnssTime", gnssTime.millis());
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void getNetworkTime(PluginCall call) {
        Clock networkTime = implementation.getNetworkTime();
        JSObject ret = new JSObject();

        if(networkTime === -1) {
            call.reject("Error getting Network Time");
        }else if(networkTime === null) {
            call.reject("Network Time is not available");
        }else{
            ret.put("networkTime", networkTime.millis());
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void getTrueTime(PluginCall call) {
        new Thread(
            () -> {
                try {
                    NTPUDPClient client = new NTPUDPClient();
                    client.setDefaultTimeout(10000);
                    client.open();
                    InetAddress hostAddr = InetAddress.getByName("pool.ntp.org");
                    TimeInfo info = client.getTime(hostAddr);
                    long returnTime = info.getMessage().getTransmitTimeStamp().getTime();

                    JSObject ret = new JSObject();
                    ret.put("networkTime", returnTime);
                    call.resolve(ret);
                } catch (Exception e) {
                    call.reject("Error getting network time", e);
                }
            }
        )
            .start();
    }
}

File: ios/Sources/CapacitorRealTimePlugin/CapacitorRealTimePlugin.swift
import Foundation
import Capacitor
import Kronos

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorRealTimePlugin)
public class CapacitorRealTimePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorRealTimePlugin"
    public let jsName = "CapacitorRealTime"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getUptime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getGnssTime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getNetworkTime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTrueTime", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapacitorRealTime()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    @objc func getUptime(_ call: CAPPluginCall) {
        let uptime = implementation.getUptime()
        call.resolve([
            "uptime": uptime
        ])
    }

    @objc func getUsedTimeZone(_ call: CAPPluginCall) {
        let timeZone = implementation.getUsedTimeZone()
        call.resolve([
            "timeZone": timeZone
        ])
    }

    @objc func getGnssTime(_ call: CAPPluginCall) {
        // Not implemented on iOS, provide an appropriate message or response
        call.unimplemented("GNSS Time is not available on iOS")
    }

    @objc func getNetworkTime(_ call: CAPPluginCall) {
        // Not implemented on iOS, provide an appropriate message or response
        call.unimplemented("Network Time is not available on iOS")
    }

    @objc func getTrueTime(_ call: CAPPluginCall) {
        Clock.sync {
            time, error in
            if let time = time {
                // time is a Date object when using Kronos
                call.resolve(["trueTime": time.timeIntervalSince1970 * 1000])
            } else if let error = error {
                // error is a TimeInterval (Double) representing the offset in seconds
                call.reject("Error getting network time. Offset: \(error) seconds")
            } else {
                call.reject("Unknown error occurred while getting network time")
            }
        }
    }
}
